gsd/gsd.c

```diff
@@ -530,199 +447,213 @@ static int gsd_is_entry_valid(struct gsd_handle* handle, size_t idx)
-                                                        * handle->header.index_allocated_entries);
-        if (handle->index == NULL)
-        {
-            return GSD_ERROR_MEMORY_ALLOCATION_FAILED;
-        }
+    return GSD_SUCCESS;
+}
 
-        bytes_read = gsd_pread_retry(handle->fd,
-                                     handle->index,
-                                     sizeof(struct gsd_index_entry)
-                                         * handle->header.index_allocated_entries,
-                                     handle->header.index_location);
+/** @internal
+    @brief Free the memory allocated by the write buffer or unmap the mapped memory.
 
-        if (bytes_read == -1
-            || bytes_read
-                   != sizeof(struct gsd_index_entry) * handle->header.index_allocated_entries)
-        {
-            return GSD_ERROR_IO;
-        }
-    }
-#if GSD_USE_MMAP
-    else if (handle->open_flags == GSD_OPEN_APPEND)
-    {
-        // in append mode, we want to avoid reading the entire index in memory, but we also don't
-        // want to bother keeping the mapping up to date. Map the index for now to determine
-        // index_num_entries, but then unmap it and use different logic to manage a cache of only
-        // unwritten index entries
-
-        // mmap may fail if offset is not a multiple of the page size, so we
-        // always memory map from the beginning of the file and then access the
-        // index pointer by its offset.
-        size_t page_size = getpagesize();
-        size_t index_size = sizeof(struct gsd_index_entry) * handle->header.index_allocated_entries;
-        size_t offset = (handle->header.index_location / page_size) * page_size;
-        handle->mapped_data = mmap(NULL,
-                                   index_size + (handle->header.index_location - offset),
-                                   PROT_READ,
-                                   MAP_SHARED,
-                                   handle->fd,
-                                   offset);
-        handle->index = (struct gsd_index_entry*)(((char*)handle->mapped_data)
-                                                  + (handle->header.index_location - offset));
-
-        if (handle->mapped_data == MAP_FAILED)
-        {
-            return GSD_ERROR_IO;
-        }
-    }
-#endif
+    @param buf Buffer to free.
 
-    // since the namelist is small, we always allocate memory for it rather than memory mapping
-    // validate that the namelist block exists inside the file
-    if (handle->header.namelist_location
-            + sizeof(struct gsd_namelist_entry) * handle->header.namelist_allocated_entries
-        > handle->file_size)
+    @returns GSD_SUCCESS on success, GSD_* error codes on error.
+*/
+static int gsd_byte_buffer_free(struct gsd_byte_buffer* buf)
+{
+    if (buf == NULL || buf->data == NULL)
     {
-        return GSD_ERROR_FILE_CORRUPT;
+        return GSD_ERROR_INVALID_ARGUMENT;
     }
```

Comment:
Do you need a separate branch for memory mapped files here?

Reply:
No, the current implementation always reads the name list directly. It needs to fully read on open anyways to hash the names for lookup.